### ISSUE TITLE
ktest --- running ktest in ktest

### ISSUE THREAD

**Issue by daejunpark** [kframework/k-legacy#533]
We need to have `ktest` be able to run `ktest` as well as `krun`.

Recently, the issues of #518, #523, and #485 reported that previously-working-features did not work any longer. As @osa1 said [https://github.com/kframework/k/issues/518#issuecomment-44646856], this is partially because of a failure of code review. However, I think this is mainly due to the _absence_ of regression test.

We do not have a regression test for `ktest` at all. We cannot have the regression test for `ktest` by using `ktest` because `ktest` cannot be executed by `ktest`.

Along with the same line of the earlier issue #396, we need to have `ktest` execute arbitrary processes other than `krun`.

**Comment by osa1** [comment]
Absence of regression tests for ktest is a problem. We already repeated lots of bugs fixed in the past.

However, while I'm not against testing ktest using ktest iteself, I don't understand how can we do that(using current implementation). Because ktest is a complex tool and testing it may need complex features. On the other hand, unit tests may give us more flexibility without any extra effort for modifying ktest(e.g. without adding that feature to ktest).

Are we going to have a "test language" that provides us all the required primitives for testing and then write "regression-533.ktest" like files for tests?(this is not testing ktest with ktest though) Or are we going to write ktest/K definition files that use something like `<ktest> ... </ktest>`?(then use ktest to read that fields and somehow run tests) (I'm not sure if that makes sense though). This sounds seriously cool but I'm not sure if that's worth the effort.

I can see four alternative implementations here, first one being the current approach but it needs adding lots of new features to current ktest:
1. Specify ktest tests in config.xml files.
2. Write a ktest language using K, and specify tests in that language.
3. Modify K to have tags like `<ktest> ... </ktest>` in definition files and specify tests using that.
4. We may skip all these and use ordinary unit tests for that.

I think it might be good to discuss how to add regression tests for some recent bugs in these alternative implementations and see advantages/disadvantages. But for now, even though I'm not an experienced K user, from an implementors point of view I can see this problems with each approach:
1. We need to modify ktest itself for lots of new features and make ktest even more complex beast. Not flexible at all.
2. I think this is really flexible and we can leave lots of things to the user, for example a user can use his rules in tests etc. In the end the user should use rules for signalling a success/failure and other than that he/she is free to use any K rules. (3) is also like that. I think only different between (2) and (3) is that in (3) we need to modify K Java source(until we bootstrap it) to add new features while in (2) we need to modify ktest definition written in K(of course we may need to modify Java parts too -- to add new primitive operations etc. but I think if we need to modify Java parts more than K parts that means K is not flexible enough yet, i.e. there are lots of things a programmer might want to do that K is not currently supporting a way to do).
3. See (2)
4. We miss the opportunity to bootstrap more K parts :-)

I think what Daejun has in mind is using `config.xml` files like we do right now to test ktest but I don't think that'll work without adding huge amounts of new features to ktest and make our config file structure 10x more complex with full of ad-hoc tags for things. Instead we can either use ordinary unit tests and have maximum flexibility with minimum cost, or redesign ktest to enable users to specify tests using a "ktest language" or `<ktest> .. </ktest>` fields in definition files.

**Comment by dwightguth** [comment]
...Not sure this is the right approach. Unit tests might be better.

**Comment by cos** [comment]
Yes, let's try to avoid complexity. Let's have a clean interface for running ktest (i.e., kompile, krun) from java/junit. Once we have that, we can expose it at the k level through either 2. or 3.

**Comment by daejunpark** [comment]
@osa1 

Omer, thank you for your valuable suggestion.

Let me explain my initial idea. I do not want to make `ktest` more complex than now; it is rather the opposite. What I want to do is to generalize `ktest` in the sense that `ktest` is not confined to run only K tools, but able to run arbitrary processes. By doing this, `ktest` will become much simpler and flexible to run arbitrary process and check their results.

Of course, I admit that this change is not simple. I'm not saying this should be done immediately, but saying that we should move to this direction in a long term. Previously, we had discussed on this topic of a re-design of ktest [https://github.com/kframework/k/pull/399#issuecomment-41816978], and my suggestion could be considered as one possible option of the re-design of ktest.

By the way, I like your idea of implementing a new feature in K. Actually, we can implement the whole of ktest in K, which was also my previous idea as you might know.

As for the unit test approach, I'm not sure how it can be easily done since I have no experience of using JUnit. For example, how can we test various `include` features including all of corner cases without writing actual `config.xml` files. I guess it will not be that simple to construct them in source code level, not easy to understand. Please excuse me and correct me if I'm totally wrong.

**Comment by dwightguth** [comment]
No see, the whole point of ktest is to be a testing framework specific to
k. If you make it generic, you might as well (and likely will instead) use
JUnit.

The way unit testing works is to abstract out dependencies and only test
small portions of the code at a  time. So you will have a test for the
config file parser which does take a  config.xml file, and then you will
have tests for each of the features in PgmArg and CmdArg and so forth.

**Comment by daejunpark** [comment]
Well, I agree that the unit testing is general enough to test anything we want. 

But I'm talking about efficiency and convenience. 

In order to do the unit testing, one should know the source code level internals of ktest. However, almost every bugs are reported by those who are not ktest developers. They cannot simply report a bug and write a regression test if the unit testing is the only way to do.

Also, even for ktest developers, I think it will take much more time to write a regression test using the unit testing approach. Even if it is not the case, we might have to re-write the unit tests whenever we change internal data structures of ktest.

I think that's why we need both the unit testing and end-to-end testing (or black-box testing). If it is not, ktest does not need to exist. Also, if ktest is needed for K tools, ktest also needs ktest because ktest is one of K tools. Furthermore, I'd like to say that we have had in mind the generalized ktest already since we started re-implement ktest2 (the current version of ktest) from scratch; well, the generalization has been postponed until now, though.

**Comment by dwightguth** [comment]
The purpose of ktest is to test k definitions, not to test the k tool
chain. We use it to test the tool chain, but it's essentially a hack and is
also a large part of why the test suite takes way too long to run.

I agree that end-to-end testing has its place in the tool, but it should be
limited in scope to those types of things that cannot be tested any other
way. We should, in the long run, have a very limited selection of ktest,
krun, kompile, and kast process invocations, and test everything that we
can using unit tests.

We already have gone considerably too far down the path of trying to use
ktest to test the tool chain when it was never the purpose it was intended
for. Now is the time to back out, not go deeper.

**Comment by traiansf** [comment]
To be frank, ktest was first designed to test the tool chain.  And then it
was improved to allow other people to use it to test their definitions.

Other than that, I fully agree with Dwight that we should try to test as
much as we can with unit tests.

On the other hand, I'm not sure there is any other way than fully testing
them to be 100% sure that all definitions in the tutorial work as expected.

2014-06-09 23:10 GMT+03:00 dwightguth notifications@github.com:

**Comment by grosu** [comment]
ktest is a tool in the K package, for users.  We also use it as developers, but let us not forget that the tools is there for our users.

I actually agree with Daejun that the ktest tool is more general than K needs.  Not sure how much we should push in this direction, but so far trying to do things generally served us well, so I am in favor in coming up with a general design.  For example, ktest really does not care whether we call kompile and krun, or javac and java, resp., I think.  So it can be viewed as a tool to test complex applications which consist of various tools and phases: compilation, execution, etc.  junit was not conceived for that.  So I see a niche here where we can come up with a general tool.

Grigore

**Comment by cos** [comment]
I will be blunt: ktest is reinventing the wheel.

Yes, it is useful now as it covers some common testing scenarios. But generalizing it to execute arbitrary processes would move it towards what build tools already do, and do very well: specifying file patterns and executing commands on them.

I think the right path is to either make ktest a special type of junit test suite with a command line shortcut, or make it into a plugin for whatever build tool we'll decide to use.

**Comment by grosu** [comment]
Well, if you put this way, then I will also put it bluntly: I think you do not understand ktest and what it tries to achieve.  You should look at its uses in the tutorial, and in the exercises repository.  If you show me how you can do that with existing wheels, as compactly and modularly as ktest does it currently, then I will listen.

**Comment by cos** [comment]
I can try. ktest has two main "tags": `<test>` and `<include>`.

`<test>` kompiles the `definition` (also, optionally, the pdf). Then, it iterates over the programs in the `programs` directory looking for files with a certain `extension` and, for each:
- `krun`s it given on optional input and some custom command line options
- compares the result against the matching file in the `results` directory

For the adventurous, there is also the option to `krun` a particular program with special command line arguments using the `<program>` tag.

```
  <test
      definition="mykool.k"
      programs="tests/examples/mykool"
      results="tests/examples/mykool"
      extension="kool" >
      <all-programs>
        <krun-option name="--color" value="off" />
      </all-programs>
      <program name="my-colorful-test.kool"> 
        <krun-option name="--color" value="on" />
      </program>
  </test>
```

`<include>` executes all the tests in the included file (and transitively). `directory` is used to rewire the directory the included config is evaluated in -- the default is the current directory. `programs` and `tests` do the same for programs and tests, relative to `directory`. `programs` and `results` from the included `<test>` will be relative to their corresponding rewired paths.

```
  <include file="../mykoolconfig.xml"
           directory="thenewroot"
           programs="tests"
           results="outputs" />
```

---

Now, for how I see this implemented using existing tools. ktest is one part configuration, one part execution. 

I would implement the execution part using JUnit. Each `kompile` and `krun` would become a JUnit test. Each config file would become a `TestSuite` --  we would likely have a config-to-`KTestSuite` convertor. There are multiple ways to implement this -- I won't go into details. 

Regarding the configuration part, xml is ok but I would go for something more succinct. I mentioned typesafe config in another thread. The `<test>` would look something like:

mykool.conf:

```
definition: ${directory}/mykool.k
programs: ${programs}/programs
results: ${results}/results
options: ${options} { "--color": off }
special {
  "my-colorful-test.kool" {
    options: ${options} { color: on }
  }
}
```

The `<include>` would look like:

```
directory: thenewroot
programs: tests
results: outputs
tests: ${tests} [ include "mycool.conf" ]
```

Essentially, mykool.conf inherits `directory`, `programs`, and `results` from its parent, and rewires them as it wants. I might not have gotten the wiring as it is in ktest, but the important aspect is that the system would be more flexible and more concise than ktest. It would also be explicit -- I like to see what happens to the paths, not to rely on some implicit knowledge.

**Comment by cos** [comment]
I don't mean to bash on ktest here. As I understand, it does its job. I'm just saying that if significant work is to be put into it, I don't think generalizing the current system is the right path.

**Comment by grosu** [comment]
I am less concerned about the implementation.  If you convert the config.xml files into something else more verbose but achieve the same result using off-the-shelf tools, for me it is equally good.  What is important here is the semantics of ktest, and not its particular implementation.  That is what I want to preserve and possibly improve.  In terms of implementation, all I want is to make sure that we do not create any dependencies: our users should only have java installed, nothing else.  This is one of the main reasons we chose Java, first place.

So it is not clear to me what you were referring to when you said that ktest is reinventing the wheel.  For me, ktest is its configuration language, which we represent using xml.  That is not reinventing anything.  It was derived from our explicit and specific needs, observing that the current tools did not give us what we wanted.  The fact that isolated users of K write their own scripts to do their own testing is not an argument against ktest, it is an argument for it.

Daejun, can you please update the "manual" with the following?

ktest is much more complex than your quick description.  The program options are necessary to everybody who writes non-trivial definitions, not only to the adventurous.  And so is the capability to exclude things.  And the programs and results can be multiple folders, not only one.  The `directory` and the `programs`/`results` have different meaning and semantics, they should never be put into the same category.  K definitions change, evolve, while the programs and their outputs stay mostly the same; sometimes they change, too, which is why you want to have the capability to override them, too.  But most of the time they don't.  For example, you have lots of programs and tests for n language features.  Now you add feature number n+1.  You want to reuse all the previous programs and their outputs, but with a different language definition.

The semantics of a ktest config file config.xml is the following pair:

```
[[config.xml]] = ( { (p1,i1,o1), (p2,i2,o2), ..., (pn,in,on) }, lang.k )
```

The first element is a set of triples (program,input,output), namely the actual tests to be performed against a K definition.  The second argument is the actual definition.  Based on this domain, we can give semantics to all ktest's constructs.  For example, `[[include config.xml]]` adds to the semantics of the current configuration file all the triples (program,input,output) from the semantics of config.xml, but ignores the definition of config.xml, because we want to test the current definition not the old one.  Other constructs, like `more-programs`, `exclude`, `programs`, etc., allow you to modify the set of triples (program,input,output) in a compact way.  This was derived by handling tens of definitions and exercises.

So:
1) You did not give me any argument for changing the design of ktest.  Since a design already exists and satisfies the current needs (modulo bugs in the implementation, which need to be fixed), if you want to change it you cannot just discard it and propose something new from scratch without giving us solid evidence that you are aware of what the current design does, to the last detail, not only in principle.  Even better, provide a prototype for an alternative.
2) I still believe that you do not understand all the desired uses of the tool.  I re-insist that you should look at all the examples in both the tutorial and the `k-exercises` repository (https://github.com/kframework/k-exercises); if you don't have access to that, let me know and I will give you access to it.  It is NOT one particular use scenario of ktest; it is a lot of uses, which essentially led to the design of the tool.
3) It is hard to believe that existing tools will give us precisely what we need.  If what you had in mind was a different way to implement ktest, then I can accept that, because for me the implementation is much more minor than the design.  However, that should not create any dependencies for our users, they should only need Java (not even javac).
4) We may want an additional include tag, maybe `import`, which would include also the definition, not only the programs/inputs/outputs.  A couple of people I talked to thought that was the meaning of `include`.  Having two different ways to import tests will at least make them reflect on the difference between the two and hopefully choose the right one for their needs.

The above being said, I am actually thankful that you go through this exercise.  It is a hard one.  However, you and anybody else should be aware that no change will be made to ktest which makes things less modular, more verbose, or less flexible.  And here we are not talking "in principle", but on more than one hundred config.xml files.  Any concrete design change proposal should start with showing how the current config.xml files, BOTH in the tutorial and in the k-exercises repository, will look like, and carefully analyze the diffs.  I bet $100 that one cannot do conceptually better than what we have now for the current case studies.  And to make it more friendly to other case studies, I think all we need is an `import` tag as described above.

**Comment by grosu** [comment]
This I agree with.  That is, if significant work needs to be put into, then we cannot justify a generalization.  Not at this stage.  But if we can get a more general tool quickly, simply by stepping back and realizing that kompile and krun are just two particular programs that need to be called in sequence and with various parameters, and that we can do the same with any other similar pairs of tools, then this may be beneficial.  Same like using category theory: makes things cleaner, simpler, and more general.

Grigore

**Comment by grosu** [comment]
I added the entire K team to the k-exercises repository.  This is likely a temporary thing, until we start teaching K again.

**Comment by daejunpark** [comment]
Regarding the recent discussion between Cosmin and Grigore,

I agree with Grigore in that re-designing/re-implementing ktest will not be convincing until it contains the generalization.

However, I also agree Cosmin that ktest seems to reinvent the wheel. Indeed, I've thought that the underlying infrastructure of ktest might exist somewhere. 

According to my experience of ktest until now, I think that the following properties are necessary for the generalized ktest:
- Execution part
  - run given processes in parallel (probably in massive parallel) with given dependencies
  - compare outputs using user-defined equivalence relation
  - report results in various formats
- Configuration part
  - composition via inheritance
  - uri pattern matching
  - parameterized with system status (such as environment variables)

Cosmin, could you please check if your suggested tools meet those criteria?

**Comment by daejunpark** [comment]
@grosu 

I created a wiki page of "Manual (to be processed)",
https://github.com/kframework/k/wiki/Manual-(to-be-processed)
and added your comment:
https://github.com/kframework/k/wiki/Manual-(to-be-processed):-KTest#ktest-internals
Note that I've added only the relevant part, and please correct it if I'm wrong.

I think it would be better to maintain this manual-to-be-processed file as a wiki page. We don't need to have it under code-review, and anyone can freely add anything that they think is worth to be processed in the K reference manual.

**Comment by grosu** [comment]
I don't think we should re-design ktest.  What we should do is:
1) Fix the bugs that were introduced
2) Keep our eyes open to user's needs and possibly add more functionality to ktest without redesigning it.  For example, we may want to have an additional `include` tag, which includes also the directory, not only the programs and results.

Grigore

**Comment by grosu** [comment]
Can you Daejun, list the current problems/bugs that need to be fixed?

Grigore

**Comment by cos** [comment]
@grosu,

Regarding semantics vs implementation, I am referring to the implementation. But yes, the implementation may mean different syntax, and this syntax is part of a language that has more expressive power than ktest.

Regarding dependencies, it would only depend on libraries that are pure Java bytecode. 

I will look at all the examples in `k-tutorial` to see if I missed anything major.

@daejunpark,

Execution: 

> - run given processes in parallel (probably in massive parallel) with given dependencies

JUnit has parallel executors. Massively parallel, probably not. But it would certainly be much easier to implement if we have a Java API to everything.

> - compare outputs using user-defined equivalence relation

yes, but, of course, it would not automatically give us an implementation for the equivalence relation itself

> - report results in various formats

JUnit returns a `TestResult` object which is interpreted by many IDEs. You could also write extra outputs for this object.

Configuration:

> - composition via inheritance

Yes, typesafe config does this. It merges configuration trees, with the latter values having priority.

> - uri pattern matching

Not by default. This is more part of the build tool's functionality than the configuration language's. Still, adding pattern matching to ktest would be cool. It would compress some of the include config file enumerations.

> - parameterized with system status (such as environment variables)

typesafe config has access to environment variables by default. Also, its Java API allows adding any other variables, or even entire configurations.

**Comment by daejunpark** [comment]
@grosu 

We already tagged 'ktest' label to ktest related issues. So, you can simply list up by selecting 'ktest' label in the issues page:
https://github.com/kframework/k/issues?labels=ktest&state=open

**Comment by daejunpark** [comment]
I close this issue since it is not interesting any longer.
